File: apps/cargo-tracker/src/contexts/tracking/domain/model/tracking-status.ts
/** 追跡状態（9 段階の追跡フェーズ。IT5 は荷役由来 5 状態、例外系は IT6） */
export const TrackingStatus = {
  NOT_RECEIVED: 'NOT_RECEIVED',
  RECEIVED: 'RECEIVED',
  LOADED: 'LOADED',
  ONBOARD_CARRIER: 'ONBOARD_CARRIER',
  UNLOADED: 'UNLOADED',
  AWAITING_CLAIM: 'AWAITING_CLAIM',
  CLAIMED: 'CLAIMED',
  EXCEPTION: 'EXCEPTION',
  UNKNOWN: 'UNKNOWN',
} as const;
export type TrackingStatus = (typeof TrackingStatus)[keyof typeof TrackingStatus];

export const TRACKING_STATUS_LABELS: Record<TrackingStatus, string> = {
  NOT_RECEIVED: '受領待ち',
  RECEIVED: '受領済',
  LOADED: '積込済',
  ONBOARD_CARRIER: '輸送中',
  UNLOADED: '荷降し済',
  AWAITING_CLAIM: '引取待ち',
  CLAIMED: '引取済',
  EXCEPTION: '例外発生',
  UNKNOWN: '不明',
};

export function isTrackingStatus(value: string): value is TrackingStatus {
  return Object.values(TrackingStatus).includes(value as TrackingStatus);
}

/**
 * 輸送フェーズの進行順序（受領前 → 引取済）。EXCEPTION / UNKNOWN は輸送フェーズ外なので順序に含めない。
 * 例外解決後の復帰先を「より進んだ状態」で決めるための比較に用いる。
 */
const TRANSPORT_PHASE_ORDER: readonly TrackingStatus[] = [
  TrackingStatus.NOT_RECEIVED,
  TrackingStatus.RECEIVED,
  TrackingStatus.LOADED,
  TrackingStatus.ONBOARD_CARRIER,
  TrackingStatus.UNLOADED,
  TrackingStatus.AWAITING_CLAIM,
  TrackingStatus.CLAIMED,
];

/**
 * 輸送フェーズの進行度（大きいほど進んでいる）。
 * EXCEPTION / UNKNOWN など順序外の状態は -1 を返す。
 */
export function transportPhaseRank(status: TrackingStatus): number {
  return TRANSPORT_PHASE_ORDER.indexOf(status);
}

File: docs/article/source/ruby/apps/cargo-tracker/packs/tracking/app/domain/tracking/domain/tracking_status.rb
# frozen_string_literal: true

module Tracking
  module Domain
    # 輸送状態（TrackingStatus）。荷役・手動更新で遷移する（IT5 対象値）。
    # 値等価の不変オブジェクト。
    class TrackingStatus
      NOT_RECEIVED = "NOT_RECEIVED"
      RECEIVED = "RECEIVED"
      LOADED = "LOADED"
      ONBOARD_CARRIER = "ONBOARD_CARRIER"
      UNLOADED = "UNLOADED"
      AWAITING_CLAIM = "AWAITING_CLAIM"
      CLAIMED = "CLAIMED"
      EXCEPTION = "EXCEPTION"

      VALUES = [
        NOT_RECEIVED, RECEIVED, LOADED, ONBOARD_CARRIER, UNLOADED, AWAITING_CLAIM, CLAIMED, EXCEPTION
      ].freeze

      attr_reader :value

      # 荷役作業種別（RECEIVE/LOAD/UNLOAD/CLAIM）に対応する輸送状態。
      FROM_HANDLING = {
        "RECEIVE" => RECEIVED, "LOAD" => LOADED, "UNLOAD" => UNLOADED, "CLAIM" => CLAIMED
      }.freeze

      def self.initial = new(value: NOT_RECEIVED)

      # 荷役種別から対応する輸送状態を返す（該当なしは nil）。
      def self.for_handling(handling_type)
        value = FROM_HANDLING[handling_type.to_s]
        value && new(value: value)
      end

      def initialize(value:)
        raise ArgumentError, "輸送状態が不正です: #{value}" unless VALUES.include?(value)

        @value = value
        freeze
      end

      def self.exception = new(value: EXCEPTION)

      def not_received? = value == NOT_RECEIVED
      def claimed? = value == CLAIMED
      def exception? = value == EXCEPTION

      def ==(other) = other.is_a?(TrackingStatus) && other.value == value
      alias eql? ==
      def hash = value.hash
      def to_s = value
    end
  end
end
